### ISSUE TITLE
MUST READ: Different way of handling redirects

### ISSUE THREAD

**Issue by Narzerus** [Narzerus/angular-permission#51]
I've noticed almost all issues reported have to do with the redirection feature since ui-router's `$stateChangeStart` has a lot of issues.

I'd love to discuss with you guys a way to avoid using $stateChangeStart for redirects and state blocking.

Any ideas? If someone comes to a solution for this I'm thinking on giving commit priviledges since they would've built an important part of this library.

**Comment by OverZealous** [comment]
My initial thoughts:

1. The main solution I can come up with, and it only solves the problem part of the time, is to allow for non-promise permissions checks.

  The idea is that you check the return of a permissions check.  If it has a `.then` property, assume it's a Promise, and use the current logic path.

  If it doesn't, then treat the result as a truthy/falsy value, and handle the response immediately (allow or cancel completely).

  This would require more work from the user, too, as they may have to have a more complex permissions checking scheme (detect if the necessary info was already loaded, then return the correct response type based on that).

2. Cache the results of a particular permissions check (probably more harmful than good), which means only the first check of a given type requires async processing.

3. Hook into `ui-router` and try to capture/store the missing properties from `$stateChangeStart`, which would be very sloppy.

I think no matter what, there needs to be a synchronous way to perform permissions checks—right now every link to a page with permissions ends up with a tiny delay.

**Comment by OverZealous** [comment]
Also, I should probably file a bug for this, but this module refires the events with the wrong values:

In `angular-permission.mdl.js`, line 30 and 39, the broadcast looks like:

```js
if (!$rootScope.$broadcast('$stateChangeStart', toState.name, toParams, fromState.name, fromParams).defaultPrevented) {
```

But the `toState` and `fromState` properties are not `strings`, but the entire objects.  It should look like this:

```js
if (!$rootScope.$broadcast('$stateChangeStart', toState, toParams, fromState, fromParams).defaultPrevented) {
```

I'm guessing it was done this way to prevent an infinite loop, but that could easily be handled via a private flag.

**Comment by qpitlove** [comment]
:+1:

**Comment by Narzerus** [comment]
Fixed in #43 keep me updated in case anything comes up.

I'll be bumping a new version in the next couple of hours.

**Comment by qpitlove** [comment]
@Narzerus :+1:
